include/aare/Cluster.hpp

```diff
@@ -1,36 +1,121 @@
+
+/************************************************
+ * @file Cluster.hpp
+ * @short definition of cluster, where CoordType (x,y) give
+ * the cluster center coordinates and data the actual cluster data
+ * cluster size is given as template parameters
+ ***********************************************/
+
 #pragma once
 
 #include <algorithm>
 #include <array>
-#include <cstddef>
 #include <cstdint>
 #include <numeric>
+#include <type_traits>
 
 namespace aare {
 
-//TODO! Template this? 
-struct Cluster3x3 {
+template <typename T, uint8_t ClusterSizeX, uint8_t ClusterSizeY,
+          typename CoordType = int16_t>
+constexpr bool is_valid_cluster =
+    std::is_arithmetic_v<T> && std::is_integral_v<CoordType> &&
+    (ClusterSizeX > 0) && (ClusterSizeY > 0);
+
+// requires clause c++20 maybe update
+template <typename T, uint8_t ClusterSizeX, uint8_t ClusterSizeY,
+          typename CoordType = int16_t,
+          typename Enable = std::enable_if_t<
+              is_valid_cluster<T, ClusterSizeX, ClusterSizeY, CoordType>>>
+struct Cluster {
```

Comment:
Maybe we can simplify the specialisations. It's a bit annoying with the repeated declaration of members.